api/layer/multipart_upload.go

```diff
@@ -201,35 +206,120 @@ func (n *layer) uploadPart(ctx context.Context, multipartInfo *data.MultipartInf
+		decSize       = p.Size
+		attributes    [][2]string
+	)
 
-	decSize := p.Size
 	if p.Info.Encryption.Enabled() {
 		r, encSize, err := encryptionReader(p.Reader, uint64(p.Size), p.Info.Encryption.Key())
 		if err != nil {
 			return nil, fmt.Errorf("failed to create ecnrypted reader: %w", err)
 		}
-		prm.Attributes = append(prm.Attributes, [2]string{AttributeDecryptedSize, strconv.FormatInt(p.Size, 10)})
-		prm.Payload = r
+		attributes = append(attributes, [2]string{AttributeDecryptedSize, strconv.FormatInt(p.Size, 10)})
+		payloadReader = r
 		p.Size = int64(encSize)
 	}
 
-	prm.Attributes[0][0], prm.Attributes[0][1] = UploadIDAttributeName, p.Info.UploadID
-	prm.Attributes[1][0], prm.Attributes[1][1] = UploadPartNumberAttributeName, strconv.Itoa(p.PartNumber)
+	var (
+		splitPreviousID      oid.ID
+		isSetSplitPreviousID bool
+		multipartHash        = sha256.New()
+		tzHash               hash.Hash
+	)
 
-	id, hash, err := n.objectPutAndHash(ctx, prm, bktInfo)
+	if n.neoFS.IsHomomorphicEnabled() {
+		tzHash = tz.New()
+	}
+
+	lastPart, err := n.treeService.GetLastPart(ctx, bktInfo, multipartInfo.ID)
 	if err != nil {
-		return nil, err
+		// if ErrPartListIsEmpty, there is the first part of multipart.
+		if !errors.Is(err, ErrPartListIsEmpty) {
+			return nil, fmt.Errorf("getLastPart: %w", err)
+		}
+	} else {
+		// try to restore hash state from the last part.
+		if binaryUnmarshaler, ok := multipartHash.(encoding.BinaryUnmarshaler); ok {
+			if err = binaryUnmarshaler.UnmarshalBinary(lastPart.MultipartHash); err != nil {
+				return nil, fmt.Errorf("unmarshalBinary: %w", err)
+			}
+		}
+
+		if n.neoFS.IsHomomorphicEnabled() {
+			if binaryUnmarshaler, ok := tzHash.(encoding.BinaryUnmarshaler); ok {
+				if err = binaryUnmarshaler.UnmarshalBinary(lastPart.HomoHash); err != nil {
+					return nil, fmt.Errorf("unmarshalBinary: %w", err)
+				}
+			}
+		}
+
+		isSetSplitPreviousID = true
+		splitPreviousID = lastPart.OID
 	}
 
+	var (
+		id           oid.ID
+		elements     []string
+		creationTime = TimeNow(ctx)
+		// User may upload part large maxObjectSize in NeoFS. From users point of view it is a single object.
+		// We have to calculate the hash from this object separately.
+		currentPartHash = sha256.New()
+	)
+
+	multipartHashes := []hash.Hash{multipartHash, currentPartHash}
+	if tzHash != nil {
+		multipartHashes = append(multipartHashes, tzHash)
+	}
+
+	prm := PrmObjectCreate{
+		Container:       bktInfo.CID,
+		Creator:         bktInfo.Owner,
+		Attributes:      attributes,
+		CreationTime:    creationTime,
+		CopiesNumber:    multipartInfo.CopiesNumber,
+		SplitID:         multipartInfo.SplitID,
+		MultipartHashes: multipartHashes,
+	}
+
+	dt := n.buffers.Get()
+	chunk := dt.(*[]byte)
+
+	// slice part manually. Simultaneously considering the part is a single object for user.
+	for {
+		if isSetSplitPreviousID {
+			prm.SplitPreviousID = &splitPreviousID
+		}
+
+		nBts, readErr := payloadReader.Read(*chunk)
+		if nBts > 0 {
```

Comment:
what if we have read lets say 10 bytes? building neofs objects that with this size? as i understand that is just a HTTP body

Reply:
so maybe reading at least some meaningful bytes number? a 10-byte object in a split for a 1TB object scares me (but TBH, i am not sure if it is possible to read HTTP body partially, i guess it is)

Reply:
If anyone is pushing 10B parts what can we do about that? Deny (but it's a valid chunk)? Cache (where?)?

Reply:
i meant we are reading http's body. if is it 10-byte long, no problem. but what if it is 60mb long but only 10 bytes were read by a reader and no EOF returned? current code creates 10-byte NeoFS object

Reply:
If you're concerned about premature `Read` (_highly_ unlikely) exit we can check `nBts` wrt upload size declared in parameters.

Reply:
Updated